### PR TITLE
[16436] - Added startup check for read_ahead_kb setting.

### PR DESCRIPTION
**Pull request by subkanthi**
[16436] - Added startup check for read_ahead_kb setting.

### CODE REVIEW

File: test/unit/org/apache/cassandra/service/StartupChecksTest.java [hunk 1]
```diff
@@ -103,6 +103,13 @@ public void compatibilityCheckIgnoresNonDbFiles() throws Exception
         startupChecks.verify();
     }
 
+    @Test
+    public void checkReadAheadKbSettingCheck() throws Exception
+    {
+        startupChecks = startupChecks.withTest(StartupChecks.checkReadAheadKbSetting);
+        startupChecks.verify();
```
Comment (subkanthi) [inline test/unit/org/apache/cassandra/service/StartupChecksTest.java:112]: Not sure if there should be some sort of assert, because this doesnt test much.

File: src/java/org/apache/cassandra/service/StartupChecks.java [hunk 2]
```diff
@@ -285,6 +286,49 @@ public void execute()
         }
     };
 
+    public static final StartupCheck checkReadAheadKbSetting = new StartupCheck()
+    {
+        // This value is in KB.
+        private static final long EXPECTED_READ_AHEAD_KB_SETTING = 8;
+        private static final String READ_AHEAD_KB_SETTING_PATH = "/sys/block/sda/queue/read_ahead_kb";
```
Comment (dpoluyanov) [inline src/java/org/apache/cassandra/service/StartupChecks.java]: Probably should be checked not only on `sda` device, but on all data directories (not sure that it's reasonable to check `read_ahead_kb` for commit log directory)
It's common to set more than one directory in `data_file_directories` property in `cassandra.yaml`
```
data_file_directories:
  - /srv/disk01
  - /srv/disk02
```

`lsblk` could be used to find mount point on common hardware (and to match device name with directory)
```
$ lsblk
NAME                         MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda                            8:0    0   1.5T  0 disk  /srv/disk01
sdb                            8:16   0   1.5T  0 disk  
`-md0                          9:0    0   5.8T  0 raid0 
  `-md0p1                    259:2    0   5.8T  0 part  /srv/disk02
...
```
but pay special attention for case with software raid, where mount point could be on raid partition like on example above
software raid also could have it's own `read_ahead_kb` and makes sense to check it too.

File: src/java/org/apache/cassandra/service/StartupChecks.java [hunk 3]
```diff
@@ -285,6 +286,49 @@ public void execute()
         }
     };
 
+    public static final StartupCheck checkReadAheadKbSetting = new StartupCheck()
+    {
+        // This value is in KB.
+        private static final long EXPECTED_READ_AHEAD_KB_SETTING = 8;
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/StartupChecks.java]: can we rename this to `MAX_RECOMMENDED_READ_AHEAD_KB_SETTING`, and set the max recommended to 128 (according to [this post](https://tobert.github.io/pages/als-cassandra-21-tuning-guide.html)) in case people use a slightly higher value for this ? We can keep the suggested value in the startup check error message to 8.

File: src/java/org/apache/cassandra/service/StartupChecks.java [hunk 4]
```diff
@@ -285,6 +286,49 @@ public void execute()
...
+                final String data = bufferedReader.readLine();
+                if (data != null)
+                {
+                    try
+                    {
+                        Integer readAheadKbSetting = Integer.parseInt(data);
+
+                        if (readAheadKbSetting > EXPECTED_READ_AHEAD_KB_SETTING)
+                        {
+                            logger.warn("High Read Ahead Kb setting: It is Recommended to set this value to 8KB " +
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/StartupChecks.java]: Maybe prefix the message with: `Detected high 'read_ahead_kb' setting for device  'sda' of data directory '/srv/disk01'.  It is Recommended to set this value to 8KB....`

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 2 above)
Comment (pauloricardomg) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: yeah ideally we'll want to go through all data dirs and dynamically find the underlying devices to be able to lookup the `read_ahead_kb` setting. I'd like to avoid using `lsblk` and use something native to java for this check if possible (ie. [Files.getFileStore](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#getFileStore(java.nio.file.Path)). Ideally we would want to do a manual test with this with multiple data dirs and a raid system like in your example to make sure the test is working. Do you have acess to a box where you can test this?

File: test/unit/org/apache/cassandra/service/StartupChecksTest.java (hunk 1 above)
Comment (pauloricardomg) [reply test/unit/org/apache/cassandra/service/StartupChecksTest.java:112]: Yeah this is ok, it's just testing the check will not throw any exception. Maybe add a little comment about this?

**Review by pauloricardomg — COMMENTED** [review]
Looking good so far. We just need to figure out the block devices dynamically based on the data dirs and test this locally.

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 2 above)
Comment (subkanthi) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: Just to clarify the steps.
1) Iterate through the data directories.
2) Get the /dev directory location using Files.getFileStore(data_directory) for the directory, Example it returns **/dev/sda2** for /home/$user/cassandra/data
3) Read the sys path ignoring the numbers from the volume returned in step 2 -  /sys/block/**sda**/queue/read_ahead_kb

Comment (dpoluyanov) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: yep, looks good, with some clarification: after step 2 collect all paths (without numbers) into `Set` before going to step 3, in order to avoid duplicate messages when two data directories points onto the same underlying device

Comment (pauloricardomg) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: sounds good 👍

File: test/unit/org/apache/cassandra/service/StartupChecksTest.java (hunk 1 above)
Comment (subkanthi) [reply test/unit/org/apache/cassandra/service/StartupChecksTest.java:112]: Added.

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 2 above)
Comment (subkanthi) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: Added logic, thanks guys.

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 3 above)
Comment (subkanthi) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: Changed to 128, thanks

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 4 above)
Comment (subkanthi) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: Changed, thanks

**Comment by subkanthi** [comment]
@pauloricardomg , @dpoluyanov I have made the changes, please check whenever u get a chance.

File: src/java/org/apache/cassandra/service/StartupChecks.java [hunk 5]
```diff
@@ -243,8 +247,8 @@ private void checkOutOfMemoryHandling()
             {
                 if (!jvmOptionsContainsOneOf("-XX:OnOutOfMemoryError="))
                     logger.warn("The JVM is not configured to stop on OutOfMemoryError which can cause data corruption."
-                            + " Either upgrade your JRE to a version greater or equal to 8u92 and use -XX:+ExitOnOutOfMemoryError/-XX:+CrashOnOutOfMemoryError"
-                            + " or use -XX:OnOutOfMemoryError=\"<cmd args>;<cmd args>\" on your current JRE.");
+                                + " Either upgrade your JRE to a version greater or equal to 8u92 and use -XX:+ExitOnOutOfMemoryError/-XX:+CrashOnOutOfMemoryError"
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/StartupChecks.java]: Please remove unrelated formatting change.

File: src/java/org/apache/cassandra/service/StartupChecks.java [hunk 6]
```diff
@@ -498,23 +573,45 @@ public void execute() throws StartupException
...
+        {
+            UntypedResultSet result = QueryProcessor.executeOnceInternal(String.format("SELECT table_name FROM %s.%s WHERE keyspace_name='%s' AND table_name='%s'",
+                                                                                       SchemaConstants.SCHEMA_KEYSPACE_NAME, "tables", SchemaConstants.AUTH_KEYSPACE_NAME, legacyAuthTable));
+            return result != null && !result.isEmpty();
+        }).collect(Collectors.toList());
 
         if (!existing.isEmpty())
             return Optional.of(String.format("Legacy auth tables %s in keyspace %s still exist and have not been properly migrated.",
-                        Joiner.on(", ").join(existing), SchemaConstants.AUTH_KEYSPACE_NAME));
+                                             Joiner.on(", ").join(existing), SchemaConstants.AUTH_KEYSPACE_NAME));
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/StartupChecks.java]: Please remove unrelated formatting change.

File: src/java/org/apache/cassandra/service/StartupChecks.java [hunk 7]
```diff
@@ -498,23 +573,45 @@ public void execute() throws StartupException
...
         List<String> existing = new ArrayList<>(SchemaConstants.LEGACY_AUTH_TABLES).stream().filter((legacyAuthTable) ->
-            {
-                UntypedResultSet result = QueryProcessor.executeOnceInternal(String.format("SELECT table_name FROM %s.%s WHERE keyspace_name='%s' AND table_name='%s'",
-                                                                                           SchemaConstants.SCHEMA_KEYSPACE_NAME,
-                                                                                           "tables",
-                                                                                           SchemaConstants.AUTH_KEYSPACE_NAME,
-                                                                                           legacyAuthTable));
-                return result != null && !result.isEmpty();
-            }).collect(Collectors.toList());
+        {
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/StartupChecks.java]: Please remove unrelated formatting change.

File: src/java/org/apache/cassandra/service/StartupChecks.java [hunk 8]
```diff
@@ -285,6 +289,77 @@ public void execute()
...
+                    if (data != null)
+                    {
+                        try
+                        {
+                            Integer readAheadKbSetting = Integer.parseInt(data);
+
+                            if (readAheadKbSetting > MAX_RECOMMENDED_READ_AHEAD_KB_SETTING)
+                            {
+                                logger.warn("Detected high 'read_ahead_kb' setting for device {} " +
+                                            "of data directory.{} It is Recommended to set this value to 8KB " +
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/StartupChecks.java]: Please fix message

File: src/java/org/apache/cassandra/service/StartupChecks.java [hunk 9]
```diff
@@ -285,6 +289,77 @@ public void execute()
...
+            for (Map.Entry<String, String> entry: blockDevices.entrySet())
+            {
+                String blockDeviceDirectory = entry.getKey();
+                String dataDirectory = entry.getValue();
+                try
+                {
+
+                    String readAheadKBPath = StartupChecks.getReadAheadKBPath(blockDeviceDirectory);
+
+                    final BufferedReader bufferedReader = Files.newBufferedReader(Path.of(readAheadKBPath));
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/StartupChecks.java]: Perhaps we can have a slightly more succint check:
```
Path readAheadKBPath = StartupChecks.getReadAheadKBPath(blockDeviceDirectory);
if (readAheadKBPath != null && Files.exists(readAheadKBPath))
{
    int readAheadKbSetting = Ints.fromBytesArray(Files.readAllBytes(readAheadKBPath));
    // perform  check
}
```
wdyt?

File: test/unit/org/apache/cassandra/service/StartupChecksTest.java [hunk 10]
```diff
@@ -103,6 +103,30 @@ public void compatibilityCheckIgnoresNonDbFiles() throws Exception
...
+    public void checkReadAheadKbSettingCheck() throws Exception
+    {
+        // This test just validates if the verify function
+        // doesn't throw any exceptions
+        startupChecks = startupChecks.withTest(StartupChecks.checkReadAheadKbSetting);
+        startupChecks.verify();
+    }
+
+    @Test
+    public void getReadAheadKBPath()
```
Comment (pauloricardomg) [inline test/unit/org/apache/cassandra/service/StartupChecksTest.java]: call this `testGetReadAheadKBPath`

File: test/unit/org/apache/cassandra/service/StartupChecksTest.java [hunk 11]
```diff
@@ -103,6 +103,30 @@ public void compatibilityCheckIgnoresNonDbFiles() throws Exception
...
+    {
+        startupChecks = startupChecks.withTest(StartupChecks.checkReadAheadKbSetting);
+
+        String sdaDirectory = startupChecks.getReadAheadKBPath("/dev/sda12");
```
Comment (pauloricardomg) [inline test/unit/org/apache/cassandra/service/StartupChecksTest.java]: use `StartupChecks.getReadAheadKBPath` to make explicit this is an static method.

File: src/java/org/apache/cassandra/service/StartupChecks.java [hunk 12]
```diff
@@ -285,6 +289,77 @@ public void execute()
...
+        // This value is in KB.
+        private static final long MAX_RECOMMENDED_READ_AHEAD_KB_SETTING = 128;
+
+        @Override
+        public void execute() throws StartupException
+        {
+            if (!FBUtilities.isLinux)
+                return;
+
+            String[] dataDirectories = DatabaseDescriptor.getRawConfig().data_file_directories;
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/StartupChecks.java]: To improve readability and testability, maybe extract this to a different method?
Something like:
```java
Map<String, String> blockDevices = getBlockDevices(DatabaseDescriptor.getRawConfig().data_file_directories);
```

Also perhaps add a simple test with a temporary directory (see: https://www.javacodegeeks.com/2021/03/creating-temporary-files-with-junit-5.html).

**Review by pauloricardomg — COMMENTED** [review]
Overall the patch looks good, added some suggestions to make the check a bit cleaner. Please let me know what do you think.

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 5 above)
Comment (subkanthi) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: Removed, thanks.

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 6 above)
Comment (subkanthi) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: Removed.

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 7 above)
Comment (subkanthi) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: Removed.

File: test/unit/org/apache/cassandra/service/StartupChecksTest.java (hunk 10 above)
Comment (subkanthi) [reply test/unit/org/apache/cassandra/service/StartupChecksTest.java]: Changed, I wasnt really sure of the convention because other test cases didnt have the test prefix.
failStartupIfInvalidSSTablesFound

File: test/unit/org/apache/cassandra/service/StartupChecksTest.java (hunk 11 above)
Comment (subkanthi) [reply test/unit/org/apache/cassandra/service/StartupChecksTest.java]: Changed. thanks.

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 9 above)
Comment (subkanthi) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: Actually not sure what is happening, changing it to below, changes the `readAheadKbSetting` value to this number `825374730` and not `128`.

```
                    int readAheadKbSetting = Ints.fromByteArray(Files.readAllBytes(readAheadKBPath));
```

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 8 above)
Comment (subkanthi) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: Sorry, whats wrong with this message

```
WARN  [main] 2022-01-02 12:26:51,241 StartupChecks.java:356 - Detected high 'read_ahead_kb' setting for device /dev/sda2 of data directory /home/parallels/Documents/GITHUB/cassandra/data/data It is Recommended to set this value to 8KB or lower as a higher value can cause high IO usage and cache churn on read-intensive workloads.
```

File: src/java/org/apache/cassandra/service/StartupChecks.java (hunk 12 above)
Comment (subkanthi) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: @pauloricardomg can you please expand on this '`Also perhaps add a simple test with a temporary directory`' , is the test for the getBlockDevices which will just return the Map?

Comment (pauloricardomg) [reply src/java/org/apache/cassandra/service/StartupChecks.java]: the idea is to verify that running `getBlockDevices(new String[]{"/tmp/mytmpdir"})` will correctly return the block device associated with this temporary directory.
